server/src/main/java/org/elasticsearch/cluster/metadata/ResettableValue.java

```diff
@@ -0,0 +1,216 @@
+        return t;
+    }
+
+    /**
+     * @return the undefined state, meaning that this value has not been specified
+     */
+    public static <T> ResettableValue<T> undefined() {
+        @SuppressWarnings("unchecked")
+        ResettableValue<T> t = (ResettableValue<T>) UNDEFINED;
+        return t;
+    }
+
+    /**
+     * Wraps a value, if the value is null, it returns {@link #undefined()}
+     */
+    public static <T> ResettableValue<T> create(T value) {
+        if (value == null) {
+            return undefined();
+        }
+        return new ResettableValue<>(true, value);
+    }
+
+    private ResettableValue(boolean isDefined, T value) {
+        this.isDefined = isDefined;
+        this.value = value;
+    }
+
+    /**
+     * @return true if the state of this is reset
+     */
+    public boolean shouldReset() {
+        return isDefined && value == null;
+    }
+
+    /**
+     * @return true when the value is defined, either with a concrete value or reset.
+     */
+    public boolean isDefined() {
+        return isDefined;
+    }
+
+    /**
+     * @return the concrete value when that's defined or null
+     */
+    public T get() {
+        return value;
+    }
+
+    /**
+     * Writes a single optional explicitly nullable value. This method is in direct relation with the
+     * {@link #read(StreamInput, Writeable.Reader)} which reads the respective value. It's the
+     * responsibility of the caller to preserve order of the fields and their backwards compatibility.
+     *
+     * @throws IOException
+     */
+    static <T> void write(StreamOutput out, ResettableValue<T> value, Writeable.Writer<T> writer) throws IOException {
+        out.writeBoolean(value.isDefined);
+        if (value.isDefined) {
+            out.writeBoolean(value.shouldReset());
+            if (value.shouldReset() == false) {
+                writer.write(out, value.get());
+            }
+        }
+    }
+
+    /**
+     * Reads a single optional and explicitly nullable value. This method is in direct relation with the
+     * {@link #write(StreamOutput, ResettableValue, Writeable.Writer)} which writes the respective value. It's the
+     * responsibility of the caller to preserve order of the fields and their backwards compatibility.
+     *
+     * @throws IOException
+     */
+    @Nullable
+    static <T> ResettableValue<T> read(StreamInput in, Writeable.Reader<T> reader) throws IOException {
+        boolean isDefined = in.readBoolean();
+        if (isDefined == false) {
+            return ResettableValue.undefined();
+        }
+        boolean shouldReset = in.readBoolean();
+        if (shouldReset) {
+            return ResettableValue.reset();
+        }
+        T value = reader.read(in);
+        return ResettableValue.create(value);
+    }
+
+    /**
+     * Gets the value and applies the function {@param f} when the value is not null.
+     */
+    public <U> U applyAndGet(Function<? super T, ? extends U> f) {
```

Comment:
Am I right that calling `applyAndGet(f)` is logically equivalent to calling `map(f).get()` (but it's more efficient 'cos it doesn't create and then throw away the wrapper object)? Maybe a name like `mapAndGet` would make that more obvious?

Reply:
Good point!